File: backend/src/entities/Log.ts
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  ManyToOne,
  JoinColumn,
} from "typeorm";
import { User } from "./User";

export enum LogLevel {
  INFO = "info",
  WARN = "warn",
  ERROR = "error",
}

@Entity("logs")
export class Log {
  @PrimaryGeneratedColumn()
  id!: number;

  @Column({ nullable: true })
  userId?: number;

  @ManyToOne(() => User, { nullable: true })
  @JoinColumn({ name: "userId" })
  user?: User;

  @Column()
  action!: string; // e.g., "login", "create_article", "delete_category"

  @Column()
  module!: string; // e.g., "auth", "articles", "categories"

  @Column()
  endpoint!: string; // e.g., "/api/v1/articles"

  @Column()
  method!: string; // GET, POST, PUT, DELETE

  @Column()
  statusCode!: number; // 200, 404, 500, etc.

  @Column({ nullable: true })
  ipAddress?: string;

  @Column({ type: "text", nullable: true })
  userAgent?: string;

  @Column("text")
  message!: string; // Mô tả chi tiết

  @Column({
    type: "varchar",
    default: LogLevel.INFO,
  })
  level!: LogLevel; // info, warn, error

  @Column("text", { nullable: true })
  metadata?: string; // JSON string lưu thêm data như request body, response, error details

  @CreateDateColumn()
  createdAt!: Date;
}

File: backend/Models/Log.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Vietsov.Api.Models;

public enum LogLevel
{
    Info,
    Warn,
    Error
}

public class Log
{
    [Key]
    public int Id { get; set; }

    public int? UserId { get; set; }

    [ForeignKey("UserId")]
    public virtual ApplicationUser? User { get; set; }

    [Required]
    [MaxLength(100)]
    public string Action { get; set; } = string.Empty;

    [Required]
    [MaxLength(50)]
    public string Module { get; set; } = string.Empty;

    [Required]
    [MaxLength(255)]
    public string Endpoint { get; set; } = string.Empty;

    [Required]
    [MaxLength(10)]
    public string Method { get; set; } = string.Empty;

    [Required]
    public int StatusCode { get; set; }

    [MaxLength(45)]
    public string? IpAddress { get; set; }

    [Column(TypeName = "text")]
    public string? UserAgent { get; set; }

    [Required]
    [Column(TypeName = "text")]
    public string Message { get; set; } = string.Empty;

    [Required]
    [MaxLength(10)]
    public LogLevel Level { get; set; } = LogLevel.Info;

    [Column(TypeName = "text")]
    public string? Metadata { get; set; } // JSON string

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
}

